cpp/ql/src/Security/CWE/CWE-416/Temporaries.qll

```diff
@@ -0,0 +1,90 @@
+import cpp
+import semmle.code.cpp.models.implementations.StdContainer
+
+/**
+ * Holds if `e` will be consumed by its parent as a glvalue and does not have
+ * an lvalue-to-rvalue conversion. This means that it will be materialized into
+ * a temporary object.
+ */
+predicate isTemporary(Expr e) {
+  e instanceof TemporaryObjectExpr
+  or
+  e.isPRValueCategory() and
+  e.getUnspecifiedType() instanceof Class and
+  not e.hasLValueToRValueConversion()
+}
+
+/** Holds if `e` is written to a container. */
+predicate isStoredInContainer(Expr e) {
+  exists(StdSequenceContainerInsert insert, Call call, int index |
+    call = insert.getACallToThisFunction() and
+    index = insert.getAValueTypeParameterIndex() and
+    call.getArgument(index) = e
+  )
+  or
+  exists(StdSequenceContainerPush push, Call call, int index |
+    call = push.getACallToThisFunction() and
+    index = push.getAValueTypeParameterIndex() and
+    call.getArgument(index) = e
+  )
+  or
+  exists(StdSequenceEmplace emplace, Call call, int index |
+    call = emplace.getACallToThisFunction() and
+    index = emplace.getAValueTypeParameterIndex() and
+    call.getArgument(index) = e
+  )
+  or
+  exists(StdSequenceEmplaceBack emplaceBack, Call call, int index |
+    call = emplaceBack.getACallToThisFunction() and
+    index = emplaceBack.getAValueTypeParameterIndex() and
+    call.getArgument(index) = e
+  )
+}
+
+/**
+ * Holds if the value of `e` outlives the enclosing full expression. For
+ * example, because the value is stored in a local variable.
+ */
+predicate outlivesFullExpr(Expr e) {
```

Comment:
I want to point out that this has changed as well as been moved so the changes need reviewing as well

Now that I think about it I should have probably moved in a separate commit to the actual changes. I don't mind doing that if you want.

Reply:
Oh, you're right. Thanks for pointing that out. I can see that there are some subtle changes here 🤔

Reply:
I think the inclusion of `not e.getConversion*().hasLValueToRValueConversion()` in all the cases is a bad idea. For instance, consider this example that we currently catch on `main`:
```cpp
const char* s = "hello";
const char* s13 = b1 ? std::string("hello").c_str() : s;
```
this is bad because of temporary `std::string` string is destroyed, but because there's an lvalue-to-rvalue conversion on the _other branch_ of the ternary we wont' flag this case any longer.

Reply:
... We should actually add this test to the `UseOfStringAfterLifetimeEnds` testcases

Reply:
I seems that the entire conditional is marked as having an `hasLValueToRValueConversion` as well as the branch `s`. But surely the conversion only needs to happen for `s`. The clang AST shows that. However our AST doesn't. Reading cppreference it suggest this is an rvalue so we don't need a a conversion here.

I can fix this by looking for this specific case but am I misunderstanding this case?

Reply:
No, I think you're understanding this correctly. I think this is case 6.1 in https://en.cppreference.com/w/cpp/language/operator_other#Conditional_operator